// @sap-ux/ui5-config/dist/middlewares.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMockServerMiddlewareConfig = exports.getFioriToolsProxyMiddlewareConfig = exports.getBackendComments = exports.getAppReloadMiddlewareConfig = void 0;
const store_1 = require("@sap-ux/store");
/**
 * Get the configuration for the AppReload middleware.
 *
 * @returns {CustomMiddleware<FioriAppReloadConfig>} the configuration
 */
function getAppReloadMiddlewareConfig() {
    return {
        name: 'fiori-tools-appreload',
        afterMiddleware: 'compression',
        configuration: {
            port: 35729,
            path: 'webapp',
            delay: 300
        }
    };
}
exports.getAppReloadMiddlewareConfig = getAppReloadMiddlewareConfig;
/**
 * Returns default comments for the given backend configuration values.
 *
 * @param backend backend config
 * @param index - optional index of backend entry
 * @returns the node comments for the backend config
 */
function getBackendComments(backend, index) {
    const comment = [];
    if (backend.authenticationType === store_1.AuthenticationType.ReentranceTicket) {
        comment.push({
            path: `configuration.backend.${index}.authenticationType`,
            comment: ' SAML support for vscode',
            key: 'authenticationType'
        });
    }
    return comment;
}
exports.getBackendComments = getBackendComments;
/**
 * Get the configuration for the Fiori tools middleware.
 *
 * @param backends configuration of backends
 * @param ui5 UI5 configuration
 * @returns {{config, comments}} configuration and comments
 */
function getFioriToolsProxyMiddlewareConfig(backends, ui5) {
    var _a, _b;
    const fioriToolsProxy = {
        name: 'fiori-tools-proxy',
        afterMiddleware: 'compression',
        configuration: {
            ignoreCertError: false
        }
    };
    let comments = [
        {
            path: 'configuration.ignoreCertError',
            comment: ' If set to true, certificate errors will be ignored. E.g. self-signed certificates will be accepted'
        }
    ];
    if (backends && backends.length > 0) {
        backends.forEach((element, index) => {
            var _a;
            element.path = (_a = element.path) !== null && _a !== void 0 ? _a : '/';
            const backendComments = getBackendComments(element, index);
            if (backendComments) {
                comments = [...comments, ...backendComments];
            }
        });
        fioriToolsProxy.configuration.backend = backends;
    }
    if (ui5 !== undefined) {
        fioriToolsProxy.configuration['ui5'] = {
            path: (_a = ui5.path) !== null && _a !== void 0 ? _a : ['/resources', '/test-resources'],
            url: (_b = ui5.url) !== null && _b !== void 0 ? _b : 'https://ui5.sap.com'
        };
        if (ui5.version) {
            fioriToolsProxy.configuration['ui5'].version = ui5.version;
        }
        if (ui5.directLoad) {
            fioriToolsProxy.configuration['ui5'].directLoad = true;
        }
    }
    return { config: fioriToolsProxy, comments };
}
exports.getFioriToolsProxyMiddlewareConfig = getFioriToolsProxyMiddlewareConfig;
const getMockServerMiddlewareConfig = (path, annotationsConfig = []) => {
    path = path === null || path === void 0 ? void 0 : path.replace(/\/$/, ''); // Mockserver is sensitive to trailing '/'
    return {
        name: 'sap-fe-mockserver',
        beforeMiddleware: 'csp',
        configuration: {
            mountPath: '/',
            services: [
                {
                    urlPath: path !== null && path !== void 0 ? path : '',
                    metadataPath: './webapp/localService/metadata.xml',
                    mockdataPath: './webapp/localService/data',
                    generateMockData: true
                }
            ],
            annotations: annotationsConfig
        }
    };
};
exports.getMockServerMiddlewareConfig = getMockServerMiddlewareConfig;
